packages/fast-components-react-base/src/viewport-positioner/viewport-positioner.tsx

```diff
@@ -0,0 +1,900 @@
+            opacity: shouldHide ? 0 : undefined,
+            position: "relative",
+            transformOrigin: `${this.state.xTransformOrigin} ${
+                this.state.yTransformOrigin
+            }`,
+            transform: `translate(
+                ${Math.floor(this.state.xTranslate)}px, 
+                ${Math.floor(this.state.yTranslate)}px
+            )`,
+            top: this.state.top === null ? null : `${this.state.top}px`,
+            right: this.state.right === null ? null : `${this.state.right}px`,
+            bottom: this.state.bottom === null ? null : `${this.state.bottom}px`,
+            left: this.state.left === null ? null : `${this.state.left}px`,
+        };
+    };
+
+    /**
+     *  Checks whether component should be disabled or not
+     */
+    private checkComponentConfig = (): void => {
+        if (!canUseDOM() || this.props.disabled === true || isNil(this.props.anchor)) {
+            this.disableComponent();
+            return;
+        }
+        this.enableComponent();
+    };
+
+    /**
+     *  Enable the component
+     */
+    private enableComponent = (): void => {
+        if (
+            !this.state.disabled ||
+            this.props.disabled ||
+            this.props.anchor === undefined ||
+            isNil(this.rootElement.current)
+        ) {
+            return;
+        }
+
+        if (
+            !(window as WindowWithIntersectionObserver).IntersectionObserver ||
+            !(window as WindowWithResizeObserver).ResizeObserver
+        ) {
+            this.setNoObserverMode();
+            return;
+        }
+
+        this.setState({
+            disabled: false,
+            noOberverMode: false,
+        });
+
+        this.collisionDetector = new (window as WindowWithIntersectionObserver).IntersectionObserver(
+            this.handleCollision,
+            {
+                root:
+                    this.props.viewport !== undefined &&
+                    this.props.viewport.current !== null
+                        ? this.props.viewport.current
+                        : null,
+                rootMargin: "0px",
+                threshold: [0, 1],
+            }
+        );
+        this.collisionDetector.observe(this.rootElement.current);
+        this.collisionDetector.observe(this.props.anchor.current);
+
+        this.resizeDetector = new (window as WindowWithResizeObserver).ResizeObserver(
+            this.handleAnchorResize
+        );
+        this.resizeDetector.observe(this.props.anchor.current);
+
+        this.props.viewport !== undefined && this.props.viewport.current !== null
+            ? this.props.viewport.current.addEventListener("scroll", this.handleScroll)
+            : document.addEventListener("scroll", this.handleScroll);
+    };
+
+    /**
+     *  Disable the component
+     */
+    private setNoObserverMode = (): void => {
+        // observers not supported so the best we do is try to set the default position if there is one
```

Comment:
I'm wondering if we really want two implementations of for subscribing to resize and intersection - I feel like it adds a good amount of code and branching logic to support both. Would we be loosing any capabilities if we omitted the observers and simply implemented the code in no observer mode?

Reply:
The resize observer has a fairly narrow use case in that it really only matters when the anchor element resizes when the anchor is offscreen and the positioner is sticking onscreen. I'd be open to getting rid of that one altogether and seeing if it becomes an issue.  The intersection observer plays a central role though - without it we'd be dependent on the height/width/scrolltop/scrolleft properties of the elements we have refs to in order to determine positioning which would not account for transforms applied to the anchor (unless we used expensive getClientBounds() calls), also, if we wanted to reposition as a result of user action we'd need to use some timer based scenario.  No observer mode basically just tries to set the default position after checking if it is a valid choice based on the axis positioning mode, doesn't even look at the spacing.

Reply:
IMO simpler more compatible version of this component would do a single positioning operation on initial mount pay the perf price for running getClientBounds at that point, and not even try to react to layout changes.